tst/unit/test_logical_location.cpp

```diff
@@ -0,0 +1,350 @@
+        idx++;
+      } while (idx < 2 * NVALID_BITS);
+    }
+
+    THEN("interleaving two zeros produces the correct bit pattern") {
+      constexpr int NVALID_BITS = 21;
+      auto interleaved = InterleaveZeros<3, NVALID_BITS>(ones);
+      std::bitset<type_size> bs_interleaved(interleaved);
+      std::cout << "Double zero interleave : " << bs_interleaved << std::endl;
+      int idx = 0;
+      do {
+        if (idx < 3 * NVALID_BITS) REQUIRE(bs_interleaved[idx] == 1);
+        idx++;
+        if (idx < 3 * NVALID_BITS) REQUIRE(bs_interleaved[idx] == 0);
+        idx++;
+        if (idx < 3 * NVALID_BITS) REQUIRE(bs_interleaved[idx] == 0);
+        idx++;
+      } while (idx < 3 * NVALID_BITS);
+    }
+  }
+}
+
+TEST_CASE("Logical Location", "[Logical Location]") {
+  GIVEN("A refinement structure") {
+    std::map<LogicalLocation, int> leaves;
+    leaves.insert({LogicalLocation(), -1});
+    RefineLocation(LogicalLocation(), leaves);
+    RefineLocation(LogicalLocation(1, 0, 0, 0), leaves);
+    RefineLocation(LogicalLocation(1, 1, 1, 1), leaves);
+    RefineLocation(LogicalLocation(2, 3, 3, 3), leaves);
+    int gid = 0;
+    for (auto &[n, leaf_id] : leaves)
+      leaf_id = gid++;
+
+    // Create a hash map of the leaves
+    std::unordered_map<LogicalLocation, int> hash_leaves;
+    hash_leaves.insert(std::begin(leaves), std::end(leaves));
+
+    // Create a set of the leaves
+    std::set<LogicalLocation> set_leaves;
+    for (const auto &[k, v] : leaves)
+      set_leaves.insert(k);
+
+    THEN("LogicalLocations store the correct Morton numbers and the map is in Morton "
+         "order") {
+      uint64_t last_morton = 0;
+      for (const auto &[leaf, leaf_id] : leaves) {
+        // Build the Morton number of this logical location by hand
+        std::bitset<64> hand_morton;
+        auto lx3 = leaf.lx3();
+        auto lx2 = leaf.lx2();
+        auto lx1 = leaf.lx1();
+        for (int i = 0; i < leaf.level(); ++i) {
+          // This is just 2^(leaf.level() - 1 - i) and we use this place by place to
+          // extract the digits of the binary representation of lx*
+          uint64_t cur_place = 1 << (leaf.level() - 1 - i);
+
+          if (lx3 / cur_place == 1) {
+            // We start at 62 because only the last 63 bits of the Morton number held
+            // in LogicalLocation store information (the leftmost bit should always be
+            // zero)
+            hand_morton[62 - (3 * i + 0)] = 1;
+          }
+          lx3 = lx3 % cur_place;
+
+          if (lx2 / cur_place == 1) {
+            hand_morton[62 - (3 * i + 1)] = 1;
+          }
+          lx2 = lx2 % cur_place;
+
+          if (lx1 / cur_place == 1) {
+            hand_morton[62 - (3 * i + 2)] = 1;
+          }
+          lx1 = lx1 % cur_place;
+        }
+        // Check that we have the correct Morton number
+        REQUIRE(hand_morton.to_ullong() == leaf.morton().bits[0]);
```

Comment:
If I understood correctly, the Morton number is split into 3 chunks with 64bit each (to encode x,y, z locations).
So why is it okay to just check one chunk here, because the others are 0 anyway? Wouldn't it be more consistent to also check for that?

Reply:
You are correct on all accounts. I am only calculating the most significant bits by hand and the rest should be zero. I will add specific requirements for that.